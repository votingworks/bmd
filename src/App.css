/* Allow selectors to be used multiple times in this file for grouping of concerns. */
/* stylelint-disable no-duplicate-selectors */

/* Reset */
html {
  box-sizing: border-box;
  line-height: 1;
  user-select: none;
}
*,
*::before,
*::after {
  box-sizing: inherit;
}

body {
  margin: 0;
}

table {
  border-collapse: collapse;
}

fieldset {
  margin: 0;
  border: none;
  padding: 0;
}

legend {
  display: block;
}

:link,
:visited {
  color: rgb(0, 0, 238);
}

:focus {
  outline: rgb(77, 144, 254) dashed 0.25rem;
}

/* Foundational */
html,
body,
#root {
  height: 100%;
  overflow: hidden;
}
@media print {
  html,
  body,
  #root {
    height: auto;
    overflow: visible;
  }
}

/* Typography */
html {
  background: #edeff0;
  color: #263238;
  font-family: -apple-system, BlinkMacSystemFont, 'Segoe UI', 'Roboto', 'Oxygen',
    'Ubuntu', 'Cantarell', 'Fira Sans', 'Droid Sans', 'Helvetica Neue',
    sans-serif;
  font-size: 24px; /* Magic number, must synced with JavaScript global defaults */
  -webkit-font-smoothing: antialiased;
  -moz-osx-font-smoothing: grayscale;
}
@media print {
  html {
    background: #ffffff;

    /* Adjust printed ballot font-size */
    font-size: 16px !important; /* stylelint-disable-line declaration-no-important */
  }
}

/* Accessibility Utilities */
.visually-hidden {
  position: absolute !important; /* stylelint-disable-line declaration-no-important */
  width: 1px;
  height: 1px;
  overflow: hidden;
  clip: rect(1px, 1px, 1px, 1px);
}
.visually-hidden a:focus,
.visually-hidden input:focus,
.visually-hidden button:focus {
  position: static;
  width: auto;
  height: auto;
}

/* Utilities */
.print-only {
  display: none;
}
@media print {
  .print-only {
    display: block;
  }
  .no-print {
    display: none;
  }
}

@media print {
  /* Chrome on Linux prints scrollbars */
  ::-webkit-scrollbar {
    display: none;
  }
}

/* stylelint-disable-next-line selector-class-pattern */
html body .cvox_indicator_container {
  /* Extra specificity/!important ensure this rule is followed. */
  display: none !important; /* stylelint-disable-line declaration-no-important */
}

/* Adjust printed ballot paper settings */
@page {
  margin: 0.375in;
  size: letter portrait;
}
