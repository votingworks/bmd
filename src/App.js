import React, { Component } from 'react';
import PrintProvider, { Print, NoPrint } from 'react-easy-print';
import './App.css';


import QrReader from 'react-qr-scanner';

import Ballot from './Ballot';
import PrintableBallot from './PrintableBallot';

var ELECTION = {
  id: "DEMO",
  title: "Demo Election",
  contests: [
    {
      id: "president",
      type: "plurality",
      name: "President",
      options: [
        "Minnie Mouse",
        "Mickey Mouse",
        "Donald Duck"
        ]
    },
    {
      id: "senator",
      type: "plurality",
      name: "Senator",
      options: [
        "Chad Hanging",
        "Lev Ermachine",
        "John Smith"
        ]
    }
  ]
};

class App extends Component {
  constructor(props) {
    super(props);
    this.state = {};
    this.election = ELECTION;
  }

  ballotReady(ballot) {
    this.setState({ballot: ballot});
  }

  print() {
    window.print();
  }

  handleScan(data) {
    if (data!=null) {
      console.log(data);
    }
  }

  handleError() {
  }

  render() {
    let mainContent, printableBallot;

    if (!this.election) {
      printableBallot = (<div></div>);

      const previewStyle = {
        height: 440,
        width: 620
      };

      mainContent = (
        <div align="center">
          <QrReader
            style={previewStyle}
            onError={this.handleError}
            onScan={this.handleScan}
          />
        </div>
      );
    } else {

      if (this.state.ballot) {
        printableBallot = (
          <PrintableBallot election={this.election} ballot={this.state.ballot} />
        );

        mainContent = (
          <div>
            <div style={{paddingLeft: "400px", paddingRight: "400px", paddingBottom: "50px"}}>{printableBallot}</div>
            <button onClick={this.print.bind(this)} style={{fontSize: "2.5em"}}>Print</button>
          </div>

        );
      } else {
        printableBallot = (<div></div>);
        mainContent = (
          <Ballot election={this.election} onBallotReady={this.ballotReady.bind(this)} />
        );
      }
    }


    return (
      <PrintProvider>
        <Print single printOnly name="ballot">
          {printableBallot}
        </Print>

        <NoPrint force>
          <div className="App">
            <img src="./vw-checkmark.png" className="App-logo" alt="logo" />
            {mainContent}
          </div>
      </NoPrint>
      </PrintProvider>
    );
  }
}

export default App;
