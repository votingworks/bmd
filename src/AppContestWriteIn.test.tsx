import React from 'react'
import { fireEvent, render, wait } from '@testing-library/react'
import fetchMock from 'fetch-mock'
import waitForExpect from 'wait-for-expect'

import App from './App'

import { noCard, voterCard, advanceTimers } from '../test/helpers/smartcards'

import {
  singleSeatContestWithWriteIn,
  setElectionInLocalStorage,
  setStateInLocalStorage,
} from '../test/helpers/election'

let currentCard = noCard
fetchMock.get('/card/read', () => JSON.stringify(currentCard))

fetchMock.post('/card/write', (url, options) => {
  currentCard = {
    present: true,
    shortValue: options.body as string,
  }
  return ''
})

jest.useFakeTimers()

beforeEach(() => {
  window.localStorage.clear()
  window.location.href = '/'
})

it(`Single Seat Contest`, async () => {
  // ====================== BEGIN CONTEST SETUP ====================== //

  setElectionInLocalStorage()
  setStateInLocalStorage()

  const { container, getByText, queryByText } = render(<App />)

  // Insert Voter Card
  currentCard = voterCard
  advanceTimers()

  // Go to Voting Instructions
  await wait(() => fireEvent.click(getByText('Get Started')))
  advanceTimers()

  // Go to First Contest
  fireEvent.click(getByText('Start Voting'))
  advanceTimers()

  // ====================== END CONTEST SETUP ====================== //

  // Advance to Single-Seat Contest with Write-In
  while (!queryByText(singleSeatContestWithWriteIn.title)) {
    fireEvent.click(getByText('Next'))
    advanceTimers()
  }

  // Test Write-In Candidate Modal Cancel
  fireEvent.click(getByText('add write-in candidate').closest('button')!)
  fireEvent.click(getByText('Cancel'))

  // Add Write-In Candidate
  fireEvent.click(getByText('add write-in candidate').closest('button')!)
  expect(getByText('Write-In Candidate')).toBeTruthy()
  // Capture styles of Single Candidate Contest
  expect(container.firstChild).toMatchSnapshot()

  // Enter Write-in Candidate Name
  fireEvent.click(getByText('B'))
  fireEvent.click(getByText('O'))
  fireEvent.click(getByText('V'))
  fireEvent.click(getByText('âŒ« delete'))
  fireEvent.click(getByText('B'))
  fireEvent.click(getByText('Accept'))
  advanceTimers()

  // Remove Write-In Candidate
  // await wait(() => getByText('BOB'))
  fireEvent.click(getByText('BOB').closest('button')!)
  fireEvent.click(getByText('Yes, Remove.'))
  advanceTimers()

  // Add Different Write-In Candidate
  fireEvent.click(getByText('add write-in candidate').closest('button')!)
  fireEvent.click(getByText('S').closest('button')!)
  fireEvent.click(getByText('A').closest('button')!)
  fireEvent.click(getByText('L').closest('button')!)
  fireEvent.click(getByText('Accept'))
  expect(getByText('SAL').closest('button')!.dataset.selected).toBe('true')

  // Try to Select Other Candidate when Max Candidates Selected.
  fireEvent.click(
    getByText(singleSeatContestWithWriteIn.candidates[0].name).closest(
      'button'
    )!
  )
  getByText(
    `You may only select ${singleSeatContestWithWriteIn.seats} candidate in this contest. To vote for ${singleSeatContestWithWriteIn.candidates[0].name}, you must first unselect selected candidate.`
  )
  fireEvent.click(getByText('Okay'))
  advanceTimers() // For 200ms Delay in closing modal

  // Go to review page and confirm write in exists
  while (!queryByText('Review Your Selections')) {
    fireEvent.click(getByText('Next'))
    advanceTimers()
  }

  // Pre-Review Screen
  fireEvent.click(getByText('Review Selections'))
  advanceTimers()

  // Review Screen
  getByText('Review Your Ballot Selections')
  expect(getByText('SAL')).toBeTruthy()
  expect(getByText('(write-in)')).toBeTruthy()

  // Print Screen
  fireEvent.click(getByText('Next'))
  advanceTimers()
  expect(getByText('Official Ballot')).toBeTruthy()
  expect(getByText('(write-in)')).toBeTruthy()
  getByText('Print your official ballot')

  // Test Print Ballot Modal
  fireEvent.click(getByText('Print Ballot'))
  fireEvent.click(getByText('Yes, print my ballot.'))
  advanceTimers()
  await waitForExpect(() => {
    expect(window.print).toBeCalled()
  })
})
