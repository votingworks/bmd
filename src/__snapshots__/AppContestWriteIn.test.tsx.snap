// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Single Seat Contest with Write In 1`] = `
.c0 {
  height: 100%;
}

.c0:focus {
  outline: none;
}

.c16 {
  border: none;
  border-radius: 0.25rem;
  box-shadow: 0 0 0 0 rgba(71,167,75,1);
  box-sizing: border-box;
  background: rgb(211,211,211);
  cursor: pointer;
  padding: 0.35rem 0.5rem;
  line-height: 1.25;
  color: black;
  touch-action: manipulation;
}

.c23 {
  border: none;
  border-radius: 0.25rem;
  box-shadow: 0 0 0 0 rgba(71,167,75,1);
  box-sizing: border-box;
  background: rgb(71,167,75);
  cursor: pointer;
  padding: 0.35rem 0.5rem;
  line-height: 1.25;
  color: #FFFFFF;
  touch-action: manipulation;
}

.c15 {
  border: none;
  border-radius: 0.25rem;
  box-shadow: 0 0 0 0 rgba(71,167,75,1);
  box-sizing: border-box;
  background: rgb(211,211,211);
  cursor: pointer;
  padding: 1rem 1.75rem;
  line-height: 1.25;
  color: black;
  font-size: 1.25rem;
  touch-action: manipulation;
}

.c22 {
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  white-space: nowrap;
}

.c22 > button {
  box-shadow: inset 1px 0 0 rgb(190,190,190);
}

.c22 > button:first-child {
  box-shadow: none;
}

.c22 > button:not(:last-child) {
  border-top-right-radius: 0;
  border-bottom-right-radius: 0;
}

.c22 > button:not(:first-child) {
  border-top-left-radius: 0;
  border-bottom-left-radius: 0;
}

.c22 > button:disabled {
  background: #028099;
  color: #ffffff;
}

.c10 {
  position: relative;
  border-radius: 0.125rem;
  box-shadow: 0 0.125rem 0.125rem 0 rgba(0,0,0,0.14), 0 0.1875rem 0.0625rem -0.125rem rgba(0,0,0,0.12), 0 0.0625rem 0.3125rem 0 rgba(0,0,0,0.2);
  background: #FFFFFF;
  cursor: pointer;
  padding: 0.5rem 0.5rem 0.5rem 4rem;
  text-align: left;
  -webkit-transition: background 0.25s,color 0.25s;
  transition: background 0.25s,color 0.25s;
}

.c10::before {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  position: absolute;
  top: 0;
  bottom: 0;
  left: 0;
  border-right: 1px solid;
  border-color: rgb(211,211,211);
  border-radius: 0.125rem 0 0 0.125rem;
  background: #FFFFFF;
  width: 3rem;
  text-align: center;
  color: #028099;
  font-size: 2rem;
  font-weight: 700;
  content: '';
}

.c2 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  overflow: auto;
}

.c4 {
  max-width: 66ch;
  line-height: 1.2;
}

.c4 h1 {
  margin: 2rem 0 1rem;
  line-height: 1.1;
  font-size: 1.5rem;
}

.c4 h2 {
  margin: 1.5rem 0 0.75rem;
  font-size: 1.25rem;
}

.c4 h3 {
  font-size: 1.17rem;
}

.c4 h4 {
  font-size: 1rem;
}

.c4 h5 {
  font-size: 0.9rem;
}

.c4 h3,
.c4 h4,
.c4 h5,
.c4 p {
  margin-top: 1rem;
  margin-bottom: 1rem;
}

.c4 h1 + h2 {
  margin-top: -0.75rem;
}

.c4 h1 + p,
.c4 h2 + p {
  margin-top: -0.75rem;
}

.c4 h3 + p,
.c4 h4 + p,
.c4 h5 + p {
  margin-top: -1rem;
}

.c4 > :first-child {
  margin-top: 0;
}

.c4 > :last-child {
  margin-bottom: 0;
}

.c4 dl {
  margin: 1rem 0;
}

.c28 {
  max-width: 66ch;
  line-height: 1.2;
}

.c28 h1 {
  margin: 2rem 0 1rem;
  line-height: 1.1;
  font-size: 1.5rem;
}

.c28 h2 {
  margin: 1.5rem 0 0.75rem;
  font-size: 1.25rem;
}

.c28 h3 {
  font-size: 1.17rem;
}

.c28 h4 {
  font-size: 1rem;
}

.c28 h5 {
  font-size: 0.9rem;
}

.c28 h3,
.c28 h4,
.c28 h5,
.c28 p {
  margin-top: 0;
  margin-bottom: 0;
}

.c28 h1 + h2 {
  margin-top: -0.75rem;
}

.c28 h1 + p,
.c28 h2 + p {
  margin-top: -0.75rem;
}

.c28 h3 + p,
.c28 h4 + p,
.c28 h5 + p {
  margin-top: 0;
}

.c28 > :first-child {
  margin-top: 0;
}

.c28 > :last-child {
  margin-bottom: 0;
}

.c28 dl {
  margin: 1rem 0;
}

.c14 {
  text-align: center;
}

.c29 {
  font-size: 0.8rem;
}

.c31 {
  font-size: 0.8rem;
  font-weight: 200;
}

.c11 {
  word-break: break-word;
}

.c30 {
  white-space: nowrap;
}

.c3 {
  margin: 0 auto;
  width: 100%;
  max-width: 35rem;
  padding: 1rem 0.75rem 0.5rem;
}

.c5 {
  text-transform: uppercase;
  font-size: 0.85rem;
  font-weight: 600;
}

.c6 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex: 1;
  -ms-flex: 1;
  flex: 1;
  position: relative;
  overflow: auto;
}

.c6::before,
.c6::after {
  position: absolute;
  z-index: 1;
  width: 100%;
  height: 0.25rem;
  content: '';
  -webkit-transition: opacity 0.25s ease;
  transition: opacity 0.25s ease;
}

.c6::before {
  top: 0;
  opacity: 0;
  background: linear-gradient( to bottom, rgb(177,186,190) 0%, transparent 100% );
}

.c6::after {
  bottom: 0;
  opacity: 0;
  background: linear-gradient( to bottom, transparent 0%, rgb(177,186,190) 100% );
}

.c7 {
  -webkit-flex: 1;
  -ms-flex: 1;
  flex: 1;
  overflow: auto;
}

.c8 {
  margin: 0 auto;
  width: 100%;
  max-width: 35rem;
  padding: 0.5rem 0.5rem 1rem;
}

.c9 {
  display: grid;
  grid-auto-rows: minmax(auto,1fr);
  grid-gap: 1.25rem;
}

.c26 {
  max-width: 250px;
}

.c27 {
  max-width: 250px;
}

.c24 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  margin: 2rem 1rem 0;
}

.c25 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  margin: auto;
}

.c25 div:first-child {
  margin-right: 1rem;
  width: 125px;
}

.c1 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  height: 100%;
}

.c1 > nav {
  -webkit-flex: 1;
  -ms-flex: 1;
  flex: 1;
}

.c1 > main {
  -webkit-flex: 2;
  -ms-flex: 2;
  flex: 2;
}

.c12 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  background-color: #333333;
  color: #ffffff;
}

.c13 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex: 1;
  -ms-flex: 1;
  flex: 1;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  padding: 1rem;
}

.c13 p > button {
  width: 100%;
}

.c17 {
  margin: 1rem 1rem 2rem;
}

.c18 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  border: 1px solid #808080;
  border-width: 1px 0;
  padding: 1rem 0;
}

.c19 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
}

.c21 button {
  min-width: 104px;
}

.c21 button[data-size='0'] {
  font-size: 18px !important;
}

.c21 button[data-size='1'] {
  font-size: 26px !important;
}

.c21 button[data-size='2'] {
  font-size: 32px !important;
}

.c20 {
  display: block;
  margin-bottom: 0.4rem;
}

@media (min-width:480px) {
  .c10 {
    padding: 1rem 1rem 1rem 4rem;
  }
}

@media (min-width:480px) {
  .c4 {
    line-height: 1.3;
  }
}

@media (min-width:480px) {
  .c28 {
    line-height: 1.3;
  }
}

@media (min-width:720px) {
  .c3 {
    padding: 1rem 1.5rem 0.5rem;
    padding-left: 5rem;
  }
}

@media (min-width:720px) {
  .c8 {
    padding-right: 1rem;
    padding-left: 1rem;
  }
}

@media print {
  .c1 {
    display: none;
  }
}

<div
  class="c0"
  tabindex="-1"
>
  <div
    class="c1"
  >
    <main
      class="c2"
    >
      <div
        class="c3"
        id="contest-header"
      >
        <div
          class="c4"
          id="audiofocus"
        >
          <h1
            aria-label="Registrar of Wills."
          >
            <div
              class="c5"
            >
              Franklin County
            </div>
            Registrar of Wills
          </h1>
          <p
            aria-label="Vote for 1. You have selected 0. Use the down arrow to hear your options. Use the right arrow to move to the next contest."
          >
            <strong>
              Vote for 
              1
              .
            </strong>
             You have selected
             
            0
            .
          </p>
        </div>
      </div>
      <div
        class="c6"
      >
        <div
          class="c7"
        >
          <div
            class="c8"
          >
            <div
              class="c9"
            >
              <button
                aria-label="Rhadka Ramachandrani, Independent."
                class="c10"
                data-choice="ramachandrani"
                data-selected="false"
                role="option"
                type="button"
              >
                <div
                  class="c4"
                >
                  <p
                    class="c11"
                  >
                    <strong>
                      Rhadka Ramachandrani
                    </strong>
                    <br />
                    Independent
                  </p>
                </div>
              </button>
              <button
                class="c10"
                data-choice="write-in"
                data-selected="false"
                role="option"
                type="button"
              >
                <div
                  class="c4"
                >
                  <p
                    aria-label="add write-in candidate."
                  >
                    <em>
                      add write-in candidate
                    </em>
                  </p>
                </div>
              </button>
            </div>
          </div>
        </div>
      </div>
    </main>
    <nav
      class="c12"
    >
      <div
        class="c13"
      >
        <div
          class="c4"
        >
          <p
            class="c14"
          >
            This is the 
            <strong>
              10th
            </strong>
             of
             
            20
             contests.
          </p>
          <p>
            <button
              class="c15"
              id="next"
              role="option"
              type="button"
            >
              Next →
            </button>
          </p>
          <p>
            <button
              class="c16"
              id="previous"
              role="option"
              type="button"
            >
              ← Back
            </button>
          </p>
        </div>
      </div>
      <div
        class="c17"
      >
        <div
          class="c18"
        >
          <div
            class="c19"
          >
            <div>
              <div
                class="c20"
              >
                Change Text Size
              </div>
              <div
                class="c21"
              >
                <span
                  class="c22"
                  data-testid="change-text-size-buttons"
                >
                  <button
                    class="c16"
                    data-size="0"
                    type="button"
                    value="0"
                  >
                    A
                  </button>
                  <button
                    class="c23"
                    data-size="1"
                    type="button"
                    value="1"
                  >
                    A
                  </button>
                  <button
                    class="c16"
                    data-size="2"
                    type="button"
                    value="2"
                  >
                    A
                  </button>
                </span>
              </div>
            </div>
          </div>
        </div>
        <div
          aria-hidden="true"
          class="c24"
          data-testid="election-info"
        >
          <div
            class="c25"
          >
            <div
              aria-hidden="true"
              class="c26"
            >
              <img
                alt="state seal"
                class="c27"
                src="/seals/Sample-Seal.svg"
              />
            </div>
            <div
              class="c28"
            >
              <h5
                aria-label=" General Election."
              >
                 General Election
              </h5>
              <p
                class="c29"
              >
                November 3, 2020
                <br />
                <span
                  class="c30"
                >
                  Franklin County
                  ,
                </span>
                 
                <span
                  class="c30"
                >
                  State of Hamilton
                </span>
              </p>
              <p
                class="c31"
              >
                <span
                  class="c30"
                >
                  Center Springfield
                  , 
                </span>
                <span
                  class="c30"
                >
                  ballot style 
                  12
                </span>
              </p>
            </div>
          </div>
        </div>
      </div>
    </nav>
  </div>
</div>
`;
